threads: free reference cycles when terminating

zerbina opened the pull request:
## Summary

When using `--gc:orc`, run a full cycle collection pass prior to exiting
the internal thread procedure, freeing all heap cells only kept alive
through reference cycles created by the thread.

Previously, they were simply not freed, with manually freeing them being
impossible once the thread terminated.

## Details

Running the cycle collector (via ``GC_fullCollect``) has to happen
*before* deallocating the ``GcThread`` instance, as the latter step
frees - when using emulated thread-local storage - the thread's
thread-local storage (which the cycle collector still depends on).

saem commented:
Very nice find, this might have been one of the hard to find leaks that was being encountered by the `cps` project.

zerbina committed:
threads: free reference cycles when terminating

Summary
=======

When using `--gc:orc`, run a full cycle collection pass prior to exiting
the internal thread procedure, freeing all heap cells only kept alive
through reference cycles created by the thread.

Previously, they were simply not freed, with manually freeing them being
impossible once the thread terminated.

Details
=======

Running the cycle collector (via ``GC_fullCollect``) has to happen
*before* deallocating the ``GcThread`` instance, as latter step frees,
when using emulated thread-local storage, the thread's thread-local
storage (which the cycle collector still depends on).

saem commented:
/merge
